datasources/src/main/java/org/opensearch/sql/datasources/storage/OpenSearchDataSourceMetadataStorage.java

```diff
@@ -72,17 +88,23 @@ public class OpenSearchDataSourceMetadataStorage implements DataSourceMetadataSt
    * @param encryptor      Encryptor.
    */
   public OpenSearchDataSourceMetadataStorage(Client client, ClusterService clusterService,
-                                             Encryptor encryptor) {
+                                             Encryptor encryptor,
+                                             List<DataSourceMetadata> keyStoreMetadataList) {
     this.client = client;
     this.clusterService = clusterService;
     this.encryptor = encryptor;
+    this.keyStoreDataSourceHashMap = new ConcurrentHashMap<>();
+    keyStoreMetadataList.forEach(metadata
+        -> keyStoreDataSourceHashMap.put(metadata.getName(), metadata));
   }
 
   @Override
   public List<DataSourceMetadata> getDataSourceMetadata() {
```

Comment:
I think that overloading a getter method is a bit confusing 
maybe create different loader method that will manage the index's lifecycle and do it eagerly from ctor ?
or maybe indicate this class has a notion of lifecycle and wrap this behaviour inside the relevant interfaces ...